Check if device is ready before taking interface up

jukkar opened the pull request:
Change the `net_if_up()` to check if the network device is ready or not. It if the device is not ready, then do not take the interface up as we cannot send or receive from the corresponding network interface anyway.

Fixes #65423

jukkar committed:
net: if: Interface stays down if device is not ready

It is pointless to take net interface up if the underlaying
device is not ready. Set also the interface status properly
in this case.

Fixes #65423

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: if: Add test to check device readiness

Add tests that check if device is ready or not, and
then verify the functionality of net_if_up() in that case.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

krish2718 commented on subsys/net/ip/net_if.c:
```diff
@@ -4360,12 +4365,23 @@ static void init_igmp(struct net_if *iface)
+	dev = net_if_get_device(iface);
+	NET_ASSERT(dev);
+
+	if (!device_is_ready(dev)) {
```
As per the linked bug this would fail as `dev->state->init_res` won't be set to failure in case device fails, its only set to failure if `services_init` fails.

jukkar replied:
Hmm, could you elaborate what is the use case because you mentioned in the issue specifically the `device_is_ready()` checks?

jukkar replied:
Coming back to this PR and I have more comments.

> Usecase1: An application that works with two network interfaces can use `device_is_ready` check to select a working interface and switch to other.

The application should not investigate the device but the network interface status if it is interested if it can send something.

> 
> Usecase2: An application that tries to bring the network interface up and implements a recovery mechanism in case if it fails multiple times, using `device_is_ready` it can skip interface up and go straight to recovery or identify that device has failed using `device_is_ready` and does chip recovery. (Of course this can be implemented in many ways, but just an example).

Here too, it is better to do things in network interface level. If the network device has implemented `start()` and `stop()`  network API functions, then the driver can return relevant status code up in the stack when calling `net_if_up()` and `net_if_down()`

rlubos commented on subsys/net/ip/net_if.c:
```diff
@@ -4295,6 +4295,11 @@ static void update_operational_state(struct net_if *iface)
 	enum net_if_oper_state prev_state = iface->if_dev->oper_state;
 	enum net_if_oper_state new_state = NET_IF_OPER_UNKNOWN;
 
+	if (!device_is_ready(net_if_get_device(iface))) {
+		new_state = NET_IF_OPER_LOWERLAYERDOWN;
+		goto exit;
+	}
+
```
Wouldn't it make sense to move this check after `net_if_is_admin_up()`? I think that if the interface was not brought up (or failed trying to do so) it should remain in `NET_IF_OPER_DOWN` state, regardless of the underlying device status. Or maybe even this part is not needed after all, since currently the inteface won't be brought ADMIN UP in case the underlying device is not ready.

jukkar replied:
Yes, I agree, I will move the check after admin up check.

jukkar commented on subsys/net/ip/net_if.c:
```diff
@@ -4360,12 +4365,23 @@ static void init_igmp(struct net_if *iface)
+	dev = net_if_get_device(iface);
+	NET_ASSERT(dev);
+
+	if (!device_is_ready(dev)) {
+		NET_DBG("Device %s (%p) is not ready", dev->name, dev);
+		status = -ENXIO;
+		goto out;
+	}
+
```
After some more investigation, I am not sure if we should actually do this. If we return error at this point and not call device driver `start()` API function, then the device does not have a chance to try to take the device up.
Or we could move this check later in this function and only check the device readiness if the L2 does not implement the `enable()`, in which case the device does not have `start()` function what we could call.
WDYT?

rlubos replied:
I think you're right, if the device is not ready, it should fail on `start()`

jukkar committed:
net: l2: dummy: Add start/stop API functions

The dummy L2 layer does not implement any L2 functionality
but it does not mean that it should not implement start/stop
functions that are called when the related network interface
is brought up or taken down.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
I moved the state change down as suggested by @rlubos, the device oper state stays the same if the device is not ready.
Also the device readiness is only checked if the device does not provide start API function as then there is nothing we can do in network level. If the device has start() implemented, then it should try to make the device ready when `net_if_up()` is called.
I also noticed that the dummy L2 did not implement enable() API which felt wrong. Dummy L2 does not provide any L2 protocol, but that does not mean that the related device should not provide start/stop functionality.

jukkar commented on subsys/net/ip/net_if.c:
```diff
@@ -4448,12 +4453,17 @@ static void init_igmp(struct net_if *iface)
 
 	net_if_lock(iface);
 
+	/* If the device is not ready it is pointless trying to take it up. */
+	dev = net_if_get_device(iface);
+	NET_ASSERT(dev);
+
```
Forgot to move this next to the `device_is_ready()` call, will send a new version that will do that.
